addons/mail_plugin/controllers/mail_plugin.py

```diff
@@ -0,0 +1,290 @@
+        # Create the partner if needed.
+        partner_info = {
+            'name': name,
+            'email': email,
+        }
+
+        #see if the partner has a parent company
+        if company and company > -1:
+            partner_info['parent_id'] = company
+        partner = request.env['res.partner'].create(partner_info)
+
+        response = {'id': partner.id}
+        return response
+
+    @http.route('/mail_plugin/log_mail_content', type="json", auth="outlook", cors="*")
+    def log_mail_content(self, model, res_id, message):
+        if model not in self._mail_content_logging_models_whitelist():
+            raise Forbidden()
+        request.env[model].browse(res_id).message_post(body=message)
+
+    def _iap_enrich(self, domain):
+        enriched_data = {}
+        try:
+            response = request.env['iap.enrich.api']._request_enrich({domain: domain})  # The key doesn't matter
+        except iap_tools.InsufficientCreditError:
+            enriched_data['enrichment_info'] = {'type': 'insufficient_credit',
+                                                'info': request.env['iap.account'].get_credits_url('reveal')}
+        except Exception:
+            enriched_data["enrichment_info"] = {'type': 'other', 'info': 'Unknown reason'}
+        else:
+            enriched_data = response.get(domain)
+            if not enriched_data:
+                enriched_data = {'enrichment_info': {'type': 'no_data',
+                                                     'info': 'The enrichment API found no data for the email provided.'}}
+        return enriched_data
+
+    def _find_existing_company(self, normalized_email):
+        sender_domain = normalized_email.split('@')[1]
+        search = sender_domain if sender_domain not in iap_tools._MAIL_DOMAIN_BLACKLIST else normalized_email
+        return request.env['res.partner'].search([('is_company', '=', True), ('email', '=ilike', '%' + search)],
+                                                 limit=1)
+
+    def _prepare_company_values(self, company):
+        if not company:
+            return {'id': -1}
+
+        return {
+                    'id': company.id,
+                    'name': company.name,
+                    'phone': company.phone,
+                    'mobile': company.mobile,
+                    'email': company.email,
+                    'image': company.image_1920,
+                    'address': {
+                        'street': company.street,
+                        'city': company.city,
+                        'zip': company.zip,
+                        'country': company.country_id.name if company.country_id else ''
+                    },
+                    'website': company.website,
+                    'additionalInfo': json.loads(company.iap_enrich_info) if company.iap_enrich_info else {}
+                }
+
+    def _create_company_from_iap(self, domain):
```

Comment:
Lots of duplicated code compared to partner_autocomplete. As we don't have enough time left, maybe add a FIXME or a NOTE to tell that it should be merged with partner_autcomplete code in 14.4+ .

Reply:
Yeah, I do think so :/ we sure have to look again at the iap related code later.